File: Common/Serialization/GenericSerializer.vb
Namespace Common

    Public MustInherit Class GenericSerializer
        Inherits Serializer

        Public Sub New(ByVal Id As String, ByVal SerializerType As Type)
            MyBase.New(Id)
            Verify.True(SerializerType.IsGenericTypeDefinition)
            Me._SerializerType = SerializerType
        End Sub

        Protected MustOverride Function GetTypeArguments(ByVal Type As Type) As Type()

        Private Function GetSerializer(ByVal TypeArguments As Type()) As Serializer
            Dim S As Serializer = Nothing
            If Me.Cache.TryGetValue(TypeArguments, S) Then
                Return S
            End If

            S = DirectCast(Me.SerializerType.MakeGenericType(TypeArguments).CreateInstance(), Serializer)

            Do Until Me.Cache.Count < CacheLimit
                Me.Cache.RemoveAt(0)
            Loop

            Me.Cache.Add(TypeArguments, S)
            Return S
        End Function

        Public NotOverridable Overrides Function CanSerializeType(ByVal Type As Type) As Boolean
            Return Me.GetTypeArguments(Type) IsNot Nothing
        End Function

        Public NotOverridable Overrides Sub [Set](Formatter As FormatterSetProxy, Obj As Object)
            Dim TypeArguments = Me.GetTypeArguments(Obj.GetType())
            Dim S = Me.GetSerializer(TypeArguments)

            Dim IsSingle = TypeArguments.Length = 1
            Formatter.Set(NameOf(IsSingle), IsSingle)

            If IsSingle Then
                Formatter.Set(NameOf(TypeArguments), TypeArguments(0))
            Else
                Formatter.Set(NameOf(TypeArguments), TypeArguments)
            End If
            S.Set(Formatter, Obj)
        End Sub

        Public NotOverridable Overrides Function [Get](Formatter As FormatterGetProxy) As Object
            Dim TypeArguments As Type()
            Dim IsSingle As Boolean
            IsSingle = Formatter.Get(Of Boolean)(NameOf(IsSingle))
            If IsSingle Then
                TypeArguments = {Formatter.Get(Of Type)(NameOf(TypeArguments))}
            Else
                TypeArguments = Formatter.Get(Of Type())(NameOf(TypeArguments))
            End If

            Dim S = Me.GetSerializer(TypeArguments)
            Return S.Get(Formatter)
        End Function

        Public NotOverridable Overrides Sub [Get](Formatter As FormatterGetProxy, Obj As Object)
            Throw New NotSupportedException()
        End Sub

#Region "SerializerType Property"
        Private ReadOnly _SerializerType As Type

        Public ReadOnly Property SerializerType As Type
            Get
                Return Me._SerializerType
            End Get
        End Property
#End Region

        Private Const CacheLimit As Integer = 10

        Private ReadOnly Cache As OrderedDictionary(Of Type(), Serializer) = New OrderedDictionary(Of Type(), Serializer)(New ArrayComparer(Of Type)())

        Public Shared Shadows Function Create(ByVal Id As String, ByVal SerializerType As Type, ByVal GetTypeArgument As Func(Of Type, Type())) As GenericSerializer
            Return New DelegateGenericSerializer(Id, SerializerType, GetTypeArgument)
        End Function

        Public Class DelegateGenericSerializer
            Inherits GenericSerializer

            Public Sub New(ByVal Id As String, ByVal SerializerType As Type, ByVal GetTypeArgument As Func(Of Type, Type()))
                MyBase.New(Id, SerializerType)
                Me.GetTypeArgumentDelegate = GetTypeArgument
            End Sub

            Protected Overrides Function GetTypeArguments(Type As Type) As Type()
                Return Me.GetTypeArgumentDelegate.Invoke(Type)
            End Function

            Private ReadOnly GetTypeArgumentDelegate As Func(Of Type, Type())

        End Class

    End Class

End Namespace

File: Common/Serialization/GenericSerializer.cs
using System;

namespace Ks.Common
{
    public abstract class GenericSerializer : Serializer
    {
        public GenericSerializer(string Id, Type SerializerType) : base(Id)
        {
            Verify.True(SerializerType.IsGenericTypeDefinition);
            this.SerializerType = SerializerType;
        }

        protected abstract Type[] GetTypeArguments(Type Type);

        private Serializer GetSerializer(Type[] TypeArguments)
        {
            if (this.Cache.TryGetValue(TypeArguments, out var S))
                return S;

            S = (Serializer) this.SerializerType.MakeGenericType(TypeArguments).CreateInstance();

            while (this.Cache.Count >= CacheLimit)
                this.Cache.RemoveAt(0);

            this.Cache.Add(TypeArguments, S);
            return S;
        }

        public sealed override bool CanSerializeType(Type Type)
        {
            return this.GetTypeArguments(Type) != null;
        }

        public sealed override void Set(FormatterSetProxy Formatter, object Obj)
        {
            var TypeArguments = this.GetTypeArguments(Obj.GetType());
            var S = this.GetSerializer(TypeArguments);

            var IsSingle = TypeArguments.Length == 1;
            Formatter.Set(nameof(IsSingle), IsSingle);

            if (IsSingle)
                Formatter.Set(nameof(TypeArguments), TypeArguments[0]);
            else
                Formatter.Set(nameof(TypeArguments), TypeArguments);
            S.Set(Formatter, Obj);
        }

        public sealed override object Get(FormatterGetProxy Formatter)
        {
            Type[] TypeArguments;
            var IsSingle = default(bool);
            IsSingle = Formatter.Get<bool>(nameof(IsSingle));
            if (IsSingle)
                TypeArguments = new[] { Formatter.Get<Type>(nameof(TypeArguments)) };
            else
                TypeArguments = Formatter.Get<Type[]>(nameof(TypeArguments));

            var S = this.GetSerializer(TypeArguments);
            return S.Get(Formatter);
        }

        public sealed override void Get(FormatterGetProxy Formatter, object Obj)
        {
            throw new NotSupportedException();
        }

        public Type SerializerType { get; }

        private const int CacheLimit = 10;

        private readonly OrderedDictionary<Type[], Serializer> Cache = new OrderedDictionary<Type[], Serializer>(new ArrayComparer<Type>());

        public static GenericSerializer Create(string Id, Type SerializerType, Func<Type, Type[]> GetTypeArgument)
        {
            return new DelegateGenericSerializer(Id, SerializerType, GetTypeArgument);
        }

        public class DelegateGenericSerializer : GenericSerializer
        {
            public DelegateGenericSerializer(string Id, Type SerializerType, Func<Type, Type[]> GetTypeArgument) : base(Id, SerializerType)
            {
                this.GetTypeArgumentDelegate = GetTypeArgument;
            }

            protected override Type[] GetTypeArguments(Type Type)
            {
                return this.GetTypeArgumentDelegate.Invoke(Type);
            }

            private readonly Func<Type, Type[]> GetTypeArgumentDelegate;
        }
    }
}
